### PR TITLE
subsys:fs:nvs Add missing test and sample documentation

### PR DESCRIPTION
**Pull request by Laczen**
NVS was missing a test for the API as well as a README.rst for the
sample. This PR corrects this.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

### CODE REVIEW

File: tests/subsys/fs/nvs/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+tests:
+  filesystem.nvs:
+    platform_whitelist: nrf51_pca10028
```
Comment (carlescufi) [inline tests/subsys/fs/nvs/testcase.yaml]: I have a feeling we should are more platforms here. What are the flash access requirements of NVS? Does it require unaligned accesses like NFFS?

Comment (Laczen) [reply tests/subsys/fs/nvs/testcase.yaml]: @carlescufi , this is the only platform I have, and the only platform I have tested it on. There are no special flash requirements from NVS, it should work on any flash. NVS dies not require unaligned access, all flash writes are sequentially (no rewrites to allready written flash locations).

Comment (nvlsianpu) [reply tests/subsys/fs/nvs/testcase.yaml]: @MaureenHelm Can you run this test on frdm_k64f, just want to know whether NVS works on the device with write-block-size of 8 bytes.

Comment (Laczen) [reply tests/subsys/fs/nvs/testcase.yaml]: @MaureenHelm, if you run the test don't forget to change the sector size to match your platform. As the test is written it uses a page size of 1kB.

Comment (carlescufi) [reply tests/subsys/fs/nvs/testcase.yaml]: @Laczen thanks for the input. I will run this on nRF52x and then you can add those to the list

**Comment by MaureenHelm** [comment]
> One thing that would make it very simple to generalize the test would be if there was a general definition of the flash erase page size

There already is. Have a look at the dts attribute `erase-block-size`, which will generate a macro `FLASH_ERASE_BLOCK_SIZE`.

**Comment by Laczen** [comment]
@Maureen Helm, thank you for the feedback on the FLASH_ERASE_BLOCK_SIZE. I
have tried it, but this doesn't work on the nrf51822.
After some search I found that this platform is probably one of the few
that does not have erase-block-size defined. It is missing in nrf51822.dtsi.

2018-05-18 18:02 GMT+02:00 Maureen Helm <notifications@github.com>:

**Comment by carlescufi** [comment]
@Laczen mind adding that in a commit as part of this PR?

**Comment by Laczen** [comment]
@MaureenHelm,Thanks for the feedback on the test. I think I have an idea
where this bus fault is coming from. Could you change in nvs.c:
line 539 and line 558: entry->len into _nvs_len_in_flash(fs, entry->len)

2018-05-18 17:55 GMT+02:00 Maureen Helm <notifications@github.com>:

**Comment by MaureenHelm** [comment]
> I think I have an idea
> where this bus fault is coming from. Could you change in nvs.c:
> line 539 and line 558: entry->len into _nvs_len_in_flash(fs, entry->len)

That didn't help.

There is also a failed assertion before the bus fault that somehow my eyes glossed over earlier:
```
starting test - nvs_test_write

    Assertion failed at /home/maureen/zephyr/tests/subsys/fs/nvs/src/main.c:59: nvs_test_write: (rc == strlen(bu)
write call failure
FAIL - nvs_test_write
```

**Comment by Laczen** [comment]
Hi @maureenhelm, Thanks for the feedback. I will see if I can order a
frdm64 board to find out what exactly goes wrong. Maybe it is some error in
the test code (my first test program).

2018-05-18 19:56 GMT+02:00 Maureen Helm <notifications@github.com>:

**Comment by Laczen** [comment]
@aurel32 or @MaureenHelm, could you run the corrected version. If it works please also give the platform it was tested on, than I can update the platform whitelist

File: tests/subsys/fs/nvs/testcase.yaml (hunk 1 above)
Comment (aurel32) [reply tests/subsys/fs/nvs/testcase.yaml]: As said in the other bug report, it partially fixes things from me, but I don't think it's the right way to fix that. The __aligned(FLASH_WRITE_BLOCK_SIZE) directive only means that the beginning of the structure has to be aligned **in RAM** to the write size. It doesn't change the size of the structure, so it means write methods are still writing things out of stack to the flash and that the stack is corrupted on reads. Also it doesn't fix writing values smaller to the block size.

In addition I am not sure using FLASH_WRITE_BLOCK_SIZE is correct here. From my understanding, it corresponding to the block size of the SoC. It doesn't necessary match the write size of an SPI flash, and we might want to use NVS on it (I don't know if it is something possible yet).

One other way to go is to change nvs_flash_write() to pad the data to write with 0, and to not use _nvs_len_in_flash for reads. I'll try to work on a patch doing that.

**Comment by Laczen** [comment]
@carlescufi, This PR can be closed, I will resubmit a PR after the changes of PR #7674 have been merged.

**Comment by carlescufi** [comment]
@Laczen I can close it, but are you sure you'd not rather update it instead of closing it and having to open a new one?

**Comment by Laczen** [comment]
@carloscufi, please close it: I am a newbie GitHub user and afraid I will
do something wrong.

**Comment by carlescufi** [comment]
Closing as per author's request

**Pull request closed without merging**